include/aws/io/uri.h

```diff
@@ -0,0 +1,122 @@
+    struct aws_byte_cursor value;
+};
+
+/**
+ * Arguments for building a URI instance. All membmers must
+ * be initialized before passing them to aws_uri_init().
+ */
+struct aws_uri_builder_args {
+    struct aws_byte_cursor scheme;
+    struct aws_byte_cursor path;
+    struct aws_byte_cursor host_name;
+    uint16_t port;
+    struct aws_array_list query_params;
+};
+
+AWS_EXTERN_C_BEGIN
+
+/**
+ * Parses 'uri_str' and initializes uri. Returns AWS_OP_SUCCESS, on success, AWS_OP_ERR on failure.
+ * After calling this function, the parts can be accessed.
+ */
+AWS_IO_API int aws_uri_init_parse(
+    struct aws_uri *uri,
+    struct aws_allocator *allocator,
+    const struct aws_byte_cursor *uri_str);
+
+/**
+ * Initializes uri to values specified in args. Returns AWS_OP_SUCCESS, on success, AWS_OP_ERR on failure.
+ * After calling this function, the parts can be accessed.
+ */
+AWS_IO_API int aws_uri_init(struct aws_uri *uri, struct aws_allocator *allocator, struct aws_uri_builder_args *args);
+AWS_IO_API void aws_uri_clean_up(struct aws_uri *uri);
+
+/**
+ * Returns the scheme portion of the uri (e.g. http, https, ftp, ftps, etc...). If the scheme was not present
+ * in the uri, the returned value will be empty. It is the users job to determine the appropriate defaults
+ * if this field is empty, based on protocol, port, etc...
+ */
+AWS_IO_API const struct aws_byte_cursor *aws_uri_scheme(const struct aws_uri *uri);
+
+/**
+ * Returns the authority portion of the uri (host[:port]). If it was not present, this was a request uri. In that
+ * case, the value will be empty.
+ */
+AWS_IO_API const struct aws_byte_cursor *aws_uri_authority(const struct aws_uri *uri);
+
+/**
+ * Returns the path portion of the uri. If the original value was empty, this value will be "/".
+ */
+AWS_IO_API const struct aws_byte_cursor *aws_uri_path(const struct aws_uri *uri);
+
+/**
+ * Returns the query string portion of the uri, minus the '?'. If not present, this value will be empty.
+ */
+AWS_IO_API const struct aws_byte_cursor *aws_uri_query_string(const struct aws_uri *uri);
+
+/**
+ * Returns the 'host_name' portion of the authority. If no authority was present, this value will be empty.
+ */
+AWS_IO_API const struct aws_byte_cursor *aws_uri_host_name(const struct aws_uri *uri);
+
+/**
+ * Returns the port portion of the authority if it was present, otherwise, returns 0.
+ * If this is 0, it is the users job to determine the correct port based on scheme and protocol.
+ */
+AWS_IO_API uint16_t aws_uri_port(const struct aws_uri *uri);
+
+/**
+ * Returns the request portion of the uri (i.e., the thing you send accross the wire).
+ */
+AWS_IO_API const struct aws_byte_cursor *aws_uri_request_uri(const struct aws_uri *uri);
+
+/**
+ * Parses query string and stores the parameters in 'params'. Returns AWS_OP_SUCCESS on success and
+ * AWS_OP_ERR on failure.
+ */
+AWS_IO_API int aws_uri_query_string_params(const struct aws_uri *uri, struct aws_array_list *params);
```

Comment:
I would name the 2nd arg`out_params`.

Also, mention to user whether array_list should be initialized or not before passing in, and mention whether or not it needs to be cleaned up after.

Why not just return a const ref to the internal datastructure? That would avoid a lot of complexity if user just wants to read values.

or just have `size_t aws_uri_query_param_count(uri)` and `param* aws_uri_query_param(uri, i)`

Reply:
I can certainly change the param name and comment updates, but there's not an internal struct for this, it's parsed on the fly.